[visionOS] Have dedicated button to enter spatial portal from element fullscreen

jyavenard opened the pull request:
#### 0d2b962e0ef808a86cf9b5336eda203e08141df6
<pre>
[visionOS] Have dedicated button to enter spatial portal from element fullscreen
<a href="https://bugs.webkit.org/show_bug.cgi?id=279171">https://bugs.webkit.org/show_bug.cgi?id=279171</a>
<a href="https://rdar.apple.com/135306130">rdar://135306130</a>

Reviewed by Jer Noble.

If the video&apos;s contain spatial metadata, display a dedicated button to go into video fullscreen
instead of using the environment picker button.
Pressing on the button will enter spatial portal mode.
The icon is for now just a generic &quot;video&quot; one.

Manually tested.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController videoControlsManagerDidChange]):
(-[WKFullScreenViewController configureEnvironmentPickerOrFullscreenVideoButtonView]):
(-[WKFullScreenViewController _removeEnvironmentPickerButtonView]):
(-[WKFullScreenViewController _removeEnvironmentFullscreenVideoButtonView]):
(-[WKFullScreenViewController loadView]):
(-[WKFullScreenViewController _enterVideoFullscreenAction:]):
(-[WKFullScreenViewController configureEnvironmentPickerButtonView]): Deleted.
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController didCleanupFullscreen]):

Canonical link: <a href="https://commits.webkit.org/283297@main">https://commits.webkit.org/283297@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f5928ccade19d5f1ea0111e947f8c6ea129e2743

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65897 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45270 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18516 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69926 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16506 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53069 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16788 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52880 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11465 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68964 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41779 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57045 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33515 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38450 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15382 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60292 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14768 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71629 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9852 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/14172 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60195 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9884 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57111 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60485 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14520 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8133 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1769 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41078 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42154 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43337 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41898 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -396,43 +406,55 @@ - (void)videoControlsManagerDidChange
-        videoPresentationInterface->setSpatialImmersive(true);
-
     LMPlayableViewController *playableViewController = videoPresentationInterface ? videoPresentationInterface->playableViewController() : nil;
     UIViewController *environmentPickerButtonViewController = playableViewController.wks_environmentPickerButtonViewController;
```
I think these local variables can be moved below the `if (mediaPlayer && [mediaPlayer spatialVideoMetadata])` block since you don't need access to these variables (and also, calling `playableViewController()` has the side effect of creating `LMPlayableViewController` earlier than strictly necessary).

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -396,43 +406,55 @@ - (void)videoControlsManagerDidChange
+        if (!playableViewController) {
+            [self _removeEnvironmentPickerOrFullscreenVideoButtonView];
+            return;
+        }
```
Is this necessary? It seems like we don't need to ensure a `LMPlayableViewController` at this point since one will be created at the point you enter video fullscreen (in `VideoPresentationInterfaceLMK::setupPlayerViewController()`).

jyavenard replied:
Yes. But we may switch from a video with metadata and one without. Like if you pause the video with metadata and mute it. It then disregard that video and pick another. In which case we will need to remove the icon and go back to the environment picker. 

And vice versa. We can only have one or another. 
Or if the last video is muted and paused we need to remove the icon whichever the type

aestes replied:
Got it. It's still unclear to me what a null (or non-null) `playableViewController` has to do with whether there is metadata.

jyavenard replied:
I should have used `videoPresentationInterface` they here indicate the same thing

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -396,43 +406,55 @@ - (void)videoControlsManagerDidChange
 
-    if (_environmentPickerButtonViewController == environmentPickerButtonViewController) {
+    RetainPtr mediaPlayer = playbackSessionInterface->linearMediaPlayer();
+    if (mediaPlayer && [mediaPlayer spatialVideoMetadata]) {
```
Nit: the null check isn't necessary (it's safe to send messages to `nil` in Objective-C and you'll get a false-y return value in that case).
```suggestion
    if ([mediaPlayer spatialVideoMetadata]) {
```

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -887,6 +916,28 @@ - (void)_togglePiPAction:(id)sender
+    if (!playbackSessionModel)
+        return;
+
+    playbackSessionModel->enterFullscreen();
```
Could you just call `WebPageProxy::enterFullscreen()` instead? That function does the same thing.

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -396,43 +406,55 @@ - (void)videoControlsManagerDidChange
+            ASSERT([[_stackView arrangedSubviews] containsObject:_enterVideoFullscreenButton.get()]);
+            return;
+        }
+        [_stackView insertArrangedSubview:_enterVideoFullscreenButton.get() atIndex:1];
```
Is it possible that there was previously an `_environmentPickerButtonViewController` that needs to be removed from the stack view before inserting `_enterVideoFullscreenButton`, say if the video changed sources while in element fullscreen?

jyavenard replied:
Yes, this is handle in the other function if we detect metadata. We then clean things up

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -396,43 +406,55 @@ - (void)videoControlsManagerDidChange
+        return;
+    }
+
+    if (_environmentPickerButtonViewController == environmentPickerButtonViewController && !_enterVideoFullscreenButtonConfigured) {
```
It's a little unfortunate that we might remove then re-add the same `_environmentPickerButtonViewController` in the case that `_enterVideoFullscreenButtonConfigured ` is true.

jyavenard replied:
That won't happen because we only have one or the other

aestes replied:
I see, you're saying if `_enterVideoFullscreenButtonConfigured` is true then `_environmentPickerButtonViewController` will be nil?

jyavenard replied:
yes. It was a bit hard to read that way. I've made the code much more explicit now.

jyavenard committed:
[visionOS] Have dedicated button to enter spatial portal from element fullscreen
https://bugs.webkit.org/show_bug.cgi?id=279171
rdar://135306130

Reviewed by Jer Noble.

If the video's contain spatial metadata, display a dedicated button to go into video fullscreen
instead of using the environment picker button.
Pressing on the button will enter spatial portal mode.
The icon is for now just a generic "video" one.

Manually tested.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController videoControlsManagerDidChange]):
(-[WKFullScreenViewController configureEnvironmentPickerOrFullscreenVideoButtonView]):
(-[WKFullScreenViewController _removeEnvironmentPickerButtonView]):
(-[WKFullScreenViewController _removeEnvironmentFullscreenVideoButtonView]):
(-[WKFullScreenViewController loadView]):
(-[WKFullScreenViewController _enterVideoFullscreenAction:]):
(-[WKFullScreenViewController configureEnvironmentPickerButtonView]): Deleted.
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController didCleanupFullscreen]):

Canonical link: https://commits.webkit.org/283297@main
